### ISSUE TITLE
runtime: goroutine stack exceeds 1000000000-byte limit

### ISSUE THREAD

**Issue by marcooriani** [rclone/rclone#8214]
#### What is the problem you are having with rclone?

I'm trying to dedupe GDrive Shared With Me

#### What is your rclone version (output from `rclone version`)

rclone v1.68.2
- os/version: darwin 15.1.1 (64 bit)
- os/kernel: 24.1.0 (arm64)
- os/type: darwin
- os/arch: arm64 (ARMv8 compatible)
- go/version: go1.23.3
- go/linking: dynamic
- go/tags: none

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

MacBook Air, Chip Apple M3, Sequoia 15.1.1

Darwin Mac.lan 24.1.0 Darwin Kernel Version 24.1.0: Thu Oct 10 21:02:26 PDT 2024; root:xnu-11215.41.3~2/RELEASE_ARM64_T8122 arm64

#### Which cloud storage system are you using? (e.g. Google Drive)

Google Drive

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

All command fails.

rclone dedupe --drive-shared-with-me --dedupe-mode rename PoiDrive:
rclone dedupe -vv --log-file bug.log --drive-shared-with-me --dedupe-mode rename PoiDrive:
rclone dedupe --drive-shared-with-me PoiDrive:

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

[Gist Bug Log](https://gist.githubusercontent.com/marcooriani/34f5baa34c93b88c271d5e72f6626726/)

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by albertony** [comment]
I haven't experience with your setup, but could it be you have symlinks creating an endless recursion during directory traversal?

**Comment by marcooriani** [comment]
I dunno if **Father Google** admit this possibility @albertony.
Anyway, let's assume it's true : 
If there is a symlinc-type recursion, i check inside Drive Shared Folder searching for shortcuts.
As you can see in thi pic, filtering for "Shortcuts" (Scorciatoie) is empty.
<img width="712" alt="Screenshot 2024-12-01 alle 08 23 14" src="https://github.com/user-attachments/assets/db056170-4e82-47f1-ab98-9188c3d5c776">
I'm pretty sure that there isn't any error about data handling, i have only a HUDGE shared folder.

**Comment by marcooriani** [comment]
I finally think is an **overload**.
I've tried to hack-me with a proxy, to debugging purpose, and i've found some interesting things.

<img width="1469" alt="Screenshot 2024-12-01 alle 09 41 19" src="https://github.com/user-attachments/assets/99d00796-9939-40d2-a2c4-6f36c0447c43">

One request, failed to load in proxy, so i repeated, and as you can see his value is insane.
**25.5 K** is fist, **5.5 K** is second, is ~ **x4.5 times.**

I dunno which type of recursive mapping RCLONE do, but probably is too huge to handle.

**Comment by ncw** [comment]
I agree with @albertony this looks like an accidental infinite recursion

```
github.com/rclone/rclone/fs/operations.dedupeDirsMap.increment(0x1400071a030, {0x140007b2bd0, 0x21})
	github.com/rclone/rclone/fs/operations/dedupe.go:276 +0xc4 fp=0x14020ce0f70 sp=0x14020ce0f30 pc=0x1053fc374
...8388470 frames elided...
```

The line in the error message is this

https://github.com/rclone/rclone/blob/bcdfad3c83720c1115c53c685bc694dc11cdac97/fs/operations/dedupe.go#L276

And here is the whole function

https://github.com/rclone/rclone/blob/bcdfad3c83720c1115c53c685bc694dc11cdac97/fs/operations/dedupe.go#L272-L278

What is happening here is rclone is recursing to find the root directory, but it never finds it as I guess there is a shortcut pointing to itself.

@marcooriani if you add this flag, does it help?

      --drive-skip-shortcuts            If set skip shortcut files

Rclone should probably be detecting recursion in shortcuts though.

**Comment by marcooriani** [comment]
**As you can see ...**

[bug.log](https://github.com/user-attachments/files/17979191/bug.log)

2024/12/02 15:16:43 DEBUG : rclone: Version "v1.68.2" starting with parameters ["rclone" "dedupe" "-vv" "--log-file" "bug.log" "--drive-skip-shortcuts" "--drive-shared-with-me" "--dedupe-mode" "rename" "PoiDrive:"]
2024/12/02 15:16:43 DEBUG : Creating backend with remote "PoiDrive:"
2024/12/02 15:16:43 DEBUG : Using config file from "/Users/marcooriani/.config/rclone/rclone.conf"
2024/12/02 15:16:43 DEBUG : PoiDrive: detected overridden config - adding "{k6zR_}" suffix to name
2024/12/02 15:16:43 DEBUG : Google drive root '': 'root_folder_id = 0AIK62lNopL0EUk9PVA' - save this in the config to speed up startup
2024/12/02 15:16:43 DEBUG : fs cache: renaming cache item "PoiDrive:" to be canonical "PoiDrive{k6zR_}:"
2024/12/02 15:16:43 INFO  : Google drive root '': Looking for duplicate names using rename mode.
runtime: goroutine stack exceeds 1000000000-byte limit
runtime: sp=0x14020c60370 stack=[0x14020c60000, 0x14040c60000]
fatal error: stack overflow

**Comment by ncw** [comment]
> **As you can see ...**
> 
> I don't tink is a infinite recursion, `runtime: goroutine stack exceeds 1000000000-byte limit`. How much byte allocated foreach cycle of `dedupeDirsMap` ???

Thanks for testing.

It is definitely an infinite recursion in rclone. That doesn't mean there is an infinite recursion in your files though.

Can you try again with `--disable ListR` please? This will use a different listing routine which is much simpler.

If that fixes the problem then there is a bug in the recursive listing routines, or quite possibly in Google's API (which has a number of bugs rclone already works around).

**Comment by marcooriani** [comment]
@ncw worked successfully !

For privacy reason, i'll include here some parts from logs :

2024/12/02 15:36:41 DEBUG : rclone: Version "v1.68.2" starting with parameters ["rclone" "dedupe" "-vv" "--log-file" "bug.log" "--disable" "ListR" "--drive-shared-with-me" "--dedupe-mode" "rename" "PoiDrive:"]
2024/12/02 15:36:41 DEBUG : Creating backend with remote "PoiDrive:"
2024/12/02 15:36:41 DEBUG : Using config file from "/Users/marcooriani/.config/rclone/rclone.conf"
2024/12/02 15:36:41 DEBUG : PoiDrive: detected overridden config - adding "{ysOIc}" suffix to name
2024/12/02 15:36:41 DEBUG : Reset feature "ListR"
2024/12/02 15:36:41 DEBUG : Google drive root '': 'root_folder_id = 0AIK62lNopL0EUk9PVA' - save this in the config to speed up startup
2024/12/02 15:36:41 DEBUG : fs cache: renaming cache item "PoiDrive:" to be canonical "PoiDrive{ysOIc}:"
2024/12/02 15:36:41 INFO  : Google drive root '': Looking for duplicate names using rename mode.
2024/12/02 15:36:52 DEBUG : pacer: Reducing sleep to 13.733583ms
2024/12/02 15:36:52 DEBUG : pacer: Reducing sleep to 40.158833ms
2024/12/02 15:36:52 DEBUG : pacer: Reducing sleep to 95.295041ms
2024/12/02 15:36:52 DEBUG : pacer: Reducing sleep to 115.609458ms
2024/12/02 15:36:52 DEBUG : pacer: Reducing sleep to 188.465458ms
2024/12/02 15:36:52 DEBUG : pacer: Reducing sleep to 257.000791ms
2024/12/02 15:36:52 DEBUG : pacer: ....
2024/12/02 15:36:52 DEBUG : pacer: ....

Other INFO / ERROR / WARNING in bug.log are with Drive Permission or skip deletion.

**Comment by ncw** [comment]
Great. I'm glad that works.

The original issue is either a bug in the rclone list recursive code or a bug in google's API - there is a history of both.

To debug further will be hard without access to the source. Can you share it with me? nick@craig-wood.com ?

If not then we can do more debugging but it will be very long winded.

Or we could say the workaround was good enough and leave well alone!

What do you think?

**Comment by marcooriani** [comment]
@ncw I sent an email, we can continue with a private conversation.
